Add pull-to-refresh (PTR) to the site picker

tonyr59h opened the pull request:
Fixes #5979.

This PR introduces a PTR view in the sites list that reloads .com+Jetpack sites in case the user has added or removed any since logging into the app. The sites are first cleared via FluxC's `REMOVE_WPCOM_AND_JETPACK_SITES` action before being fetched.

To see it in action take I've recorded two short clips:

[The first](https://cldup.com/EG4q-Fhffc.webm) demonstrates refreshing after adding a site on the web.

[The second](https://cldup.com/7HNxzafJwl.webm) demonstrates refreshing after removing a site.

To test:
1. Login with a .com account and load up the site picker
2. Add a new site on the web (Switch Site -> Add new site)
3. Once the new site is created PTR in the site picker

The list should refresh with the new site added. Now test removing it and refreshing:
1. Remove the site on the web (Site Settings -> General -> Delete your site permanently)
2. Once the site has been removed PTR in the site picker

The list should refresh with the site removed.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerActivity.java:
```diff
@@ -217,13 +222,46 @@ public void onSiteRemoved(OnSiteRemoved event) {
+                            ToastUtils.showToast(SitePickerActivity.this, R.string.site_picker_no_network_refresh_error);
+                            return;
+                        }
+                        mDispatcher.dispatch(SiteActionBuilder.newRemoveWpcomAndJetpackSitesAction());
```
Hey @tonyr59h , got to say, this all-sites removal makes me a bit uncomfortable. I assume the reason you added it is to "clean up" any remotely-deleted sites, right? Is there some reason beyond that?

Here's the thing though: an unadvertised removal like this can possibly delete data not synced to the backend yet, like draft posts that have local changes. Fixing/catering for a case like that might easily fall outside the scope of this PR so, may I suggest instead to avoid removing local sites en-mass? Perhaps we can remove local sites after we get the updated remote list and check that the site doesn't have any local drafts. WDYT?

aforcier replied:
I agree that we probably shouldn't do a mass removal here - and I wanted to add that it would be better to handle clearing any remotely-removed sites in FluxC instead (this way we'll also have the same behavior here and for the existing 15 minute auto-refresh of the list).

So perhaps after a `FETCH_SITES` completes, we track WP.com sites linked to that account that weren't in the new fetch, and remove them after we finish updating the db (all in FluxC).

tonyr59h replied:
> I assume the reason you added it is to "clean up" any remotely-deleted sites, right? Is there some reason beyond that?

That's exactly the reason. The response to `FETCH_SITES` doesn't tell us which sites were fetched so figuring out which sites should be removed is not possible here.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerActivity.java:
```diff
@@ -92,6 +96,7 @@ public void onCreate(Bundle savedInstanceState) {
         restoreSavedInstanceState(savedInstanceState);
         setupActionBar();
         setupRecycleView();
+        initSwipeToRefreshHelper(findViewById(android.R.id.content));
```
I noticed that the refresh indicator is not preserved when rotating the device. On slower connections that could be rather confusing so, let's add support for rotation, wdyt? Thanks!

tonyr59h replied:
Most definitely will fix this. My refreshes finish almost instantly so I never saw this happen :/

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerActivity.java:
```diff
@@ -217,13 +222,46 @@ public void onSiteRemoved(OnSiteRemoved event) {
+                        }
+                        if (!NetworkUtils.checkConnection(SitePickerActivity.this)) {
+                            mSwipeToRefreshHelper.setRefreshing(false);
+                            ToastUtils.showToast(SitePickerActivity.this, R.string.site_picker_no_network_refresh_error);
```
`NetworkUtils .checkConnection()` already pops up a Toast about the missing network connection so, the extra Toast is kinda unneeded. Let's remove the extra Toast, wdyt? Alternatively, you can directly use `NetworkUtils.isNetworkAvailable()` that doesn't show a Toast on its own but, I think `checkConnection` should be enough. Thanks!

tonyr59h replied:
Removed the extra toast and corresponding error string in 22286aaad0d2a969438446e2ac06eeebab3205a2

tonyr59h commented:
I've removed the site deletion call and started [a branch](https://github.com/wordpress-mobile/WordPress-FluxC-Android/tree/issue/remove-deleted-sites-on-fetch) on FluxC to take care of that particular problem.

Rotation is handled gracefully as of 10d7040f6badabeee1aab857822398a57da4c6d6, tested by delaying the network call for 5s so I could rotate.

That addresses all the feedback. Thanks for the review @hypest, think we're good for another look!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerActivity.java:
```diff
@@ -256,8 +256,6 @@ public void onRefreshStarted() {
                             mSwipeToRefreshHelper.setRefreshing(false);
                             return;
                         }
-                        mDispatcher.dispatch(SiteActionBuilder.newRemoveWpcomAndJetpackSitesAction());
```
With the removal of the sites Remove, the `onSiteRemoved()` callback seems like dead code, isn't it? I'd suggest deleting it and we can add it perhaps later if needed. Thanks!

tonyr59h replied:
`onSiteRemoved` should still be called when a self-hosted site is removed. I'll double-check and get back to you on that.

tonyr59h replied:
Just checked and it does get hit after removing a self-hosted site 👍

hypest replied:
Just curious, would that be a Jetpack site? I wonder how can a self-hosted site can be remotely removed while we still are on the SitePicker. Thanks!

tonyr59h replied:
Yes, it's a Jetpack site.

tonyr59h commented:
> Let's open an issue about this so to track it.

👍

tonyr59h commented:
Opened #6065 to track error handling for the pending FluxC changes.

hypest commented:
:shipit:
